core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -561,6 +593,234 @@ public UpdateNode plan(Update node)
+
+        ImmutableList.Builder<WhenClause> whenClauses = ImmutableList.builder();
+        for (int caseNumber = 0; caseNumber < merge.getMergeCases().size(); caseNumber++) {
+            MergeCase mergeCase = merge.getMergeCases().get(caseNumber);
+
+            if (mergeCase.getExpression().isPresent()) {
+                subPlan = subqueryPlanner.handleSubqueries(subPlan, mergeCase.getExpression().get(), analysis.getSubqueries(mergeCase));
+            }
+
+            ImmutableList.Builder<Expression> rowBuilder = ImmutableList.builder();
+            List<ColumnHandle> mergeCaseSetColumns = mergeCaseColumnsHandles.get(caseNumber);
+            for (ColumnHandle dataColumnHandle : mergeAnalysis.getDataColumnHandles()) {
+                int index = mergeCaseSetColumns.indexOf(dataColumnHandle);
+                if (index >= 0) {
+                    Expression setExpression = mergeCase.getSetExpressions().get(index);
+                    Type type = analysis.getCoercion(setExpression);
+                    if (type != null) {
+                        setExpression = new Cast(setExpression, toSqlType(type));
+                    }
+                    rowBuilder.add(subPlan.rewrite(setExpression));
+                }
+                else {
+                    Integer fieldNumber = requireNonNull(mergeAnalysis.getColumnHandleFieldNumbers().get(dataColumnHandle), "Field number for ColumnHandle is null");
+                    rowBuilder.add(target.getFieldMappings().get(fieldNumber).toSymbolReference());
+                }
+            }
+
+            // Build the match condition for the MERGE case
+            MergeCaseKind mergeKind = getMergeCaseKind(mergeCase);
+
+            // Add a boolean column which is true if a target table row was matched
+            rowBuilder.add(new IsNotNullPredicate(presentColumn.toSymbolReference()));
+            // Add the operation number
+            rowBuilder.add(new GenericLiteral("INTEGER", String.valueOf(mergeKind.getOperationNumber())));
+
+            Optional<Expression> rewritten = mergeCase.getExpression().map(subPlan::rewrite);
+            Expression condition = presentColumn.toSymbolReference();
+            if (mergeCase instanceof MergeInsert) {
+                condition = new IsNullPredicate(presentColumn.toSymbolReference());
+            }
+
+            if (rewritten.isPresent()) {
+                condition = ExpressionUtils.and(condition, rewritten.get());
+            }
+
+            whenClauses.add(new WhenClause(condition, new Row(rowBuilder.build())));
+        }
+
+        ImmutableList.Builder<Expression> rowBuilder = ImmutableList.builder();
+        mergeAnalysis.getDataColumnSchemas().forEach(columnSchema ->
+                rowBuilder.add(new Cast(new NullLiteral(), toSqlType(columnSchema.getType()))));
+        rowBuilder.add(new IsNotNullPredicate(presentColumn.toSymbolReference()));
```

Comment:
It is a boolean, and I've fixed the documentation.  Also, I fixed the bug that it was being read in the engine as if it were an int array.  It's now read using the BOOLEAN type rather than accessing the underlying array directly.